src/Common/ThreadStatus.h

```diff
@@ -57,45 +57,57 @@ using ThreadStatusPtr = ThreadStatus *;
   * Create via CurrentThread::initializeQuery (for queries) or directly (for various background tasks).
   * Use via CurrentThread::getGroup.
   */
+class ThreadGroupStatus;
+using ThreadGroupStatusPtr = std::shared_ptr<ThreadGroupStatus>;
+
 class ThreadGroupStatus
 {
 public:
-    struct ProfileEventsCountersAndMemory
-    {
-        ProfileEvents::Counters::Snapshot counters;
-        Int64 memory_usage;
-        UInt64 thread_id;
-    };
-
-    mutable std::mutex mutex;
+    /// The first thread created this thread group
+    UInt64 master_thread_id = 0;
 
     ProfileEvents::Counters performance_counters{VariableContext::Process};
     MemoryTracker memory_tracker{VariableContext::Process};
 
-    ContextWeakPtr query_context;
-    ContextWeakPtr global_context;
+    /// Access to the members below has to be in critical section with  mutex
+    mutable std::mutex mutex;
 
     InternalTextLogsQueueWeakPtr logs_queue_ptr;
     InternalProfileEventsQueueWeakPtr profile_queue_ptr;
-    std::function<void()> fatal_error_callback;
-
-    std::unordered_set<UInt64> thread_ids;
-    std::unordered_set<ThreadStatusPtr> threads;
-
-    /// The first thread created this thread group
-    UInt64 master_thread_id = 0;
 
     LogsLevel client_logs_level = LogsLevel::none;
 
     String query;
     UInt64 normalized_query_hash = 0;
 
-    std::vector<ProfileEventsCountersAndMemory> finished_threads_counters_memory;
+    /// When new query starts, new thread group is created for it, current thread becomes master thread of the query
+    static ThreadGroupStatusPtr createForQuery(ContextPtr query_context_, std::function<void()> fatal_error_callback_ = {});
 
-    std::vector<ProfileEventsCountersAndMemory> getProfileEventsCountersAndMemoryForThreads();
-};
+    const std::vector<UInt64> getInvolvedThreadIds() const;
 
-using ThreadGroupStatusPtr = std::shared_ptr<ThreadGroupStatus>;
+    void link(ThreadStatusPtr thread);
+    void unlink(ThreadStatusPtr thread);
+
+    ContextWeakPtr getQueryContextWeak() const;
+    ContextWeakPtr getGlobalContextWeak() const;
+
+    using FatalErrorCallback = std::function<void()>;
+    FatalErrorCallback getFatalErrorCallback() const;
+
+private:
+    /// Set up at creation, no race when reading
+    ContextWeakPtr query_context;
+    ContextWeakPtr global_context;
+
+    /// Set up at creation, no race when reading
+    FatalErrorCallback fatal_error_callback;
+
+    /// Set of all thread ids which has been attached to the group
+    std::unordered_set<UInt64> thread_ids;
+
+    /// Set of active threads
+    std::unordered_set<ThreadStatusPtr> threads;
```

Comment:
That's a bit funny that `thread_ids` are for all threads and `threads` are only for active ones.
btw now I can't find any place where `threads` is used. Looks like before it was used only in `ThreadGroupStatus::getProfileEventsCountersAndMemoryForThreads`, which is removed now

Reply:
Will remove it then!